src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,1376 @@
+			d_list_del_init(&f_shard->fs_list);
+			remap_add_one(remap_list, f_shard);
+
+			/* Continue with the failed shard has minimal fseq */
+			if (current == remap_list) {
+				current = &f_shard->fs_list;
+			} else {
+				f_tmp = d_list_entry(current,
+						     struct failed_shard,
+						     fs_list);
+				if (f_shard->fs_fseq < f_tmp->fs_fseq)
+					current = &f_shard->fs_list;
+			}
+			continue; /* try next spare */
+		}
+next_fail:
+		fail_count = 0;
+		if (spare_avail) {
+			/* The selected spare target is up and ready */
+			l_shard->po_target = spare_tgt->ta_comp.co_id;
+			l_shard->po_fseq = f_shard->fs_fseq;
+
+			/*
+			* Mark the shard as 'rebuilding' so that read will
+			* skip this shard.
+			*/
+			if (f_shard->fs_status == PO_COMP_ST_DOWN) {
+				l_shard->po_rebuilding = 1;
+				f_shard->fs_tgt_id = spare_tgt->ta_comp.co_id;
+			}
+		} else {
+			l_shard->po_shard = -1;
+			l_shard->po_target = -1;
+		}
+		current = current->next;
+	}
+
+	mapless_remap_dump(remap_list, md, "after remap:");
+	return 0;
+}
+
+static int
+mapless_obj_spec_place_get(struct pl_mapless_map *mmap, daos_obj_id_t oid,
+			   struct pool_target **target, uint8_t *dom_used,
+			   uint32_t dom_bytes)
+{
+	struct pool_target      *tgts;
+	unsigned int            tgts_nr;
+	d_rank_t                rank;
+	int                     tgt;
+	unsigned int            pos;
+	int rc;
+
+	D_ASSERT(daos_obj_id2class(oid) == DAOS_OC_R3S_SPEC_RANK ||
+		 daos_obj_id2class(oid) == DAOS_OC_R1S_SPEC_RANK ||
+		 daos_obj_id2class(oid) == DAOS_OC_R2S_SPEC_RANK);
+
+	/* locate rank in the pool map targets */
+	tgts = pool_map_targets(mmap->mmp_map.pl_poolmap);
+	tgts_nr = pool_map_target_nr(mmap->mmp_map.pl_poolmap);
+	/* locate rank in the pool map targets */
+	rank = daos_oclass_sr_get_rank(oid);
+	tgt = daos_oclass_st_get_tgt(oid);
+
+	for (pos = 0; pos < tgts_nr; pos++) {
+		if (rank == tgts[pos].ta_comp.co_rank &&
+		    (tgt == tgts[pos].ta_comp.co_index))
+			break;
+	}
+	if (pos == tgts_nr)
+		return -DER_INVAL;
+
+	*target = &(tgts[pos]);
+
+
+	struct pool_domain *current_dom;
+	struct pool_domain *root;
+
+	rc = pool_map_find_domain(mmap->mmp_map.pl_poolmap, PO_COMP_TP_ROOT,
+				  PO_COMP_ID_ALL, &root);
+	if (rc == 0) {
+		D_ERROR("Could not find root node in pool map.");
+		return -DER_NONEXIST;
+	}
+	current_dom = root;
+
+	while (current_dom->do_children != NULL) {
+		struct pool_domain *temp_dom;
+		int index;
+		uint64_t child_pos;
+
+		child_pos = (current_dom->do_children) - root;
+
+		for (index = 0; index < current_dom->do_child_nr; ++index) {
+			temp_dom = &(current_dom->do_children[index]);
+			int num_children = temp_dom->do_target_nr;
+			int last = num_children - 1;
+			struct pool_target *start = &(temp_dom->do_targets[0]);
+			struct pool_target *end = &(temp_dom->do_targets[last]);
+
+			if ((start <= (*target)) && ((*target) <= end)) {
+				current_dom = temp_dom;
+				setbit(dom_used, (index + child_pos));
+				break;
+			}
```

Comment:
Are you referring to this code?

> if ((start <= (*target)) && ((*target) <= end)) {
> 				current_dom = temp_dom;
> 				setbit(dom_used, (index + child_pos));
> 				break;
> 			}

The other places I do this calculate the start and ending differently as well as do different things if the target is located between the two points.
I'm not sure what I would factor out that would be meaningful.

Reply:
This one "recurses" down the fault domains and sets those domains as used so the other domains will be selected before this shards top level domain is reused. This is necessary because the SPEC shard isn't placed organically using the jump consistent hash algorithm. 

The loop in get_rebuild_target() only looks at the targets in the bottom level domain when picking a target to guarantee multiple shards are not placed on the same target. 

The loops look a little similar but one is only checking top level domains that contains a single target and the other is only checking the target level domain for ANY target in the current layout. I'm not sure they could be combined in any meaningful way.